src/stream.h

```diff
@@ -55,6 +55,21 @@ GIT_INLINE(ssize_t) git_stream_write(git_stream *st, const char *data, size_t le
 	return st->write(st, data, len, flags);
 }
 
+GIT_INLINE(int) git_stream__write_full(git_stream *st, const char *data, size_t len, int flags)
```

Comment:
Two questions:

1. Should we not just make `git_stream_write` always loop and write everything it can?  I'm not sure I see the value of having both functions.
2. I worry a bit that we should guard against the underlying write returning `0`.  I trust something like `write(2)` to not return `0` when in blocking mode, but we've got a few different implementations and things like the stransport implementation are apparently already expecting us to work in a way that they don't document (from reading that comment in its write function).  I wouldn't want to get into a place where we get in an infinite loop when write returns `0` for some reason we didn't expect.  Maybe convert it to an internal error or assert?

Reply:
1. Thing is, `git_stream_write` is used as a primitive in e.g. our mbedtls and openssl streams, where it implements the BIO callbacks. Looping is handled by mbedtls and openssl, so I wasn't quite sure whether we really want to loop ourselves again. Also, we cannot really change `git_stream_write` to loop itself, as it gets passed an `size_t` length but returns `ssize_t`. Thus, we either loop and return a wrong value, have to raise an error if `len > SSIZE_MAX` or we need to break the API.

2. Yeah, we probably should. I've made it return an error now

Reply:
Ah, yes.  I remember thinking about this however briefly before throwing up my hands and realizing that I wasn't going to fix anything quickly, so I just went to bed.  I agree with this solution for #1.